File: 2DVirtualWorldSimulatorCpp/2DVirtualWorldSimulator/Sosnowsky_hogweed.cpp
#include <iostream>
#include "Sosnowsky_hogweed.h"
#include "World.h" 

Sosnowsky_hogweed::Sosnowsky_hogweed() {
	this->name = 'h';
	this->strength = 10;
}

void Sosnowsky_hogweed::action(World* world) {
	//Kills every animal in its immediate neighborhood except cyber - sheep.
	if (position.x - 1 >= 0 ) {
		killAnimalAround(world, position.x - 1, position.y);
	}
	if (position.x + 1 < world->GetHorizSize()) {
		killAnimalAround(world, position.x + 1, position.y);
	}
	if (position.y - 1 >= 0) {
		killAnimalAround(world, position.x, position.y - 1);
	}
	if (position.y + 1 < world->GetVerticSize()) {
		killAnimalAround(world, position.x, position.y + 1);
	}
	this->SetOrganismWantedPosition(position.x, position.y);
}

void Sosnowsky_hogweed::killAnimalAround(World* world, int posX, int posY) {
	if (world->getOrganismVectorData(posY, posX) != nullptr && (world->getOrganismVectorData(posY, posX)->GetName() > LETTER_RANGE_ANIMAL_A
		&& world->getOrganismVectorData(posY, posX)->GetName() < LETTER_RANGE_ANIMAL_Z)) {
		world->getOrganismVectorData(posY, posX)->SetOrganismLifespan(false);

		world->addCommentatorVector(KILL, this, world->getOrganismVectorData(posY, posX), NULL);
		world->setOrganismVectorData(nullptr, posY, posX);
	}
}

bool Sosnowsky_hogweed::collision(World* world,Organism* organismToFight) {
	//Kills any animal which eats it, apart from cyber - sheep.
	pos animalPos;
	if (this->isAttacker == true) {
		return true;
	}
	else {
		world->addCommentatorVector(CONSUMPTION, organismToFight, this, NULL);
		world->addCommentatorVector(KILL, this, organismToFight, NULL);
		organismToFight->SetOrganismLifespan(false);
		this->isAlive = false;
		animalPos = organismToFight->GetOrganismPosition();
		world->setOrganismVectorData(nullptr, animalPos.y, animalPos.x);
		world->setOrganismVectorData(nullptr, position.y, position.x);
	}
	return true;
}

char Sosnowsky_hogweed::GetName() const {
	return 'h';
}

string Sosnowsky_hogweed::GetOrganismFullName() const {
	return "Sosnowsky_hogweed";
}

Sosnowsky_hogweed::~Sosnowsky_hogweed() {

}

File: 2D_sym_python/plants/Sosnowsky_hogweed.py
from Plant import*

class Sosnowsky_hogweed(Plant):

    def __init__(self):
        super().__init__()
        self.name='h'
        self.strength=10
    
    def action(self, world):
        if self.position['x']-1 >= 0:
            self.killAnimalAround(world, self.position['x']-1,self.position['y'])
        if self.position['x']+1 < world.GetHorizSize():
            self.killAnimalAround(world, self.position['x']+1,self.position['y'])
        if self.position['y']-1 >= 0:
            self.killAnimalAround(world, self.position['x'],self.position['y']-1)
        if self.position['y']+1 < world.GetVerticSize():
            self.killAnimalAround(world, self.position['x'],self.position['y']+1)
        self.SetOrganismWantedPosition(self.position['x'],self.position['y'])
    
    def killAnimalAround(self, world, posX, posY):
        if world.GetOrganismVectorData(posY,posX) != None and (LETTER_RANGE_ANIMAL_A < ord(world.GetOrganismVectorData(posY,posX).GetName()) < LETTER_RANGE_ANIMAL_Z) and world.GetOrganismVectorData(posY,posX).GetName()!='C':
            world.GetOrganismVectorData(posY,posX).SetOrganismLifespan(False)

            world.AddCommentatorVector(reportType.KILL, self,world.GetOrganismVectorData(posY,posX),0)
            world.SetOrganismVectorData(None,posY,posX)
    
    def collision(self, world, organismToFight):
        if self.isAttacker == True:
            return True
        else:
            for i in range(world.GetSizeHogweedList()):
                if (self.position['x']==world.GetHogweedList(i).position['x']) and (self.position['y']==world.GetHogweedList(i).position['y']):
                    world.RemoveFromHogweedList(i)
                    break

            world.AddCommentatorVector(reportType.CONSUMPTION, organismToFight, self, 0)
            world.AddCommentatorVector(reportType.KILL, self, organismToFight, 0)
            organismToFight.SetOrganismLifespan(False)
            self.isAlive = False
            animalPos = organismToFight.GetOrganismPosition()
            world.SetOrganismVectorData(None, animalPos['y'], animalPos['x'])
            world.SetOrganismVectorData(None, self.position['y'], self.position['x'])
        return True
    
    def GetName(self):
        return 'h'
    
    def GetOrganismFullName(self):
        return "Sosnowsky_hogweed"